pkg/upload/path_provider.go

```diff
@@ -0,0 +1,117 @@
+package upload
+
+import (
+	"path"
+	"sync"
+	"time"
+
+	"github.com/rs/xid"
+)
+
+// PathProvider provides path by request to upload data. The provided path is built from '<prefix>/<partition>/<unique id>'.
+// The partition is updated every interval or when the number of IDs generated passes the specified size.
+type PathProvider struct {
+	mu  sync.Mutex
+	cfg *PathProviderConfig
+	// timestamp used to calculate the time based ID value
+	timestamp time.Time
+	// count make sure we will not produce more than 'count' IDs
+	count int
+	// partition current set and reset by size and time
+	partition string
+}
+
+const (
+	DefaultDataSize     = 50000
+	DefaultDataInterval = 5 * time.Minute
+	DefaultDataPrefix   = "data"
+
+	// unixYear4000 epoch value for Saturday, January 1, 4000 12:00:00 AM
+	unixYear4000 = 64060588800
+)
+
+type TimeNow func() time.Time
+
+var (
+	DefaultPathProvider = NewPathProvider()
+	DefaultTimeNow      = time.Now
+)
+
+type PathProviderConfig struct {
+	Size     int
+	Interval time.Duration
+	Prefix   string
+	TellTime TimeNow
+}
+
+type PathProviderOption func(o *PathProviderConfig)
+
+func WithPathProviderSize(size int) PathProviderOption {
+	return func(cfg *PathProviderConfig) {
+		cfg.Size = size
+	}
+}
+
+func WithPathProviderInterval(d time.Duration) PathProviderOption {
+	return func(cfg *PathProviderConfig) {
+		cfg.Interval = d
+	}
+}
+
+func WithPathProviderPrefix(p string) PathProviderOption {
+	return func(cfg *PathProviderConfig) {
+		cfg.Prefix = p
```

Comment:
Prefix needs sanitization

Reply:
can you elaborate? this is a prefix in object store not local filesystem.

Reply:
Right - what if I pass `data/` instead of `data`?